site/_posts/2024-11-27-fish-4.md

```diff
@@ -0,0 +1,78 @@
+
+Community is most important. Open-source projects live and die based on the community, on people volunteering their time. Rust has cultivated an active, enthusiastic community, with a systems programming bent. This community offers fish-shell a strong network of potential contributors.
+
+Rust's underlying semantics are also similar to C++, which allowed for a straightforward, incremental port. We used (forked) versions of [CXX](https://cxx.rs) and [autocxx](https://google.github.io/autocxx/) to allow C++ and Rust to interoperate. Had we chosen a language with (say) green threads or garbage collection, it would have complicated the FFI, and likely forced a redesign of fish's signal handling and async-signal-safe code.
+
+But the killer feature of Rust, from fish-shell's perspective, is Send and Sync, statically enforcing rules around threading. "Fearless concurrency" is too strong - you can still blow your leg off with fork or signal handlers - but Send and Sync will be the key to unlocking fully multithreaded execution, with confidence in its correctness.
+
+### The port
+
+The port went more or less according to plan. Some new contributors pitched in, as we hoped - thank you to [Xiretza](https://github.com/Xiretza) and [henrikhorluck](https://github.com/henrikhorluck).
+
+Some parts of fish proved too entangled to port separately, resulting in some large PRs. And some bugs were introduced, and then later discovered and fixed. But overall the plan was highly successful: fish is now in Rust!
+
+### Hurdles
+
+Rust is powerful and widely appreciated, but we discovered areas of immaturity or missing features:
+
+- Configuration-time checks. For example, `cur_term` is a _symbol_ on some Linux distributions but a C _macro_ on others, making it difficult to consistently access in Rust. We wrote [our own crate](https://github.com/mqudsi/rsconf) to support configuration-time checks.
+
+- Universal macOS binaries, a surprise given Rust's reputation for cross-compilation. We implemented this manually with a shell script.
+
+- Runtime format strings, which are needed for localization. We used gettext and (our own) printf instead of using `format!`.
+
+- libc macros: `_PATH_BSHELL`, `MNT_LOCAL`, `MB_CUR_MAX`, etc. We exposed these via a C file, which (sadly) requires a C compiler.
+
+- We [implemented printf](https://crates.io/crates/fish-printf), as the existing 'printf' crates in crates.io were not standards-conforming and/or required nightly Rust.
+
+- Auxiliary tasks such as building Sphinx documentation, installer packages, running tests not written in Rust, etc. Cargo is focused exclusively on Rust so we needed another tool to drive these tasks, and we kept CMake.
+
+The big surprise from the port was retaining CMake for these auxiliary tasks. fish can be built and run via cargo alone, but CMake still drives the rest. We have a newfound respect for CMake, which has solved many of the problems we encountered, even as we transition off of it.
```

Comment:
I think this line and the one above are saying more or less the same thing? "The big surprise" is a good sentence and could be put in front of the bullet point "Auxiliary tasks"

Reply:
We could also go into a bit more detail for people who aren't as familiar as we are with the codebase.

"tests not written in Rust" could get an explanation - "fish has a big test suite consisting of not just unit tests in rust, but also integration tests as .fish script files and interactive tests via pexpect".

Also we have 1004 completion scripts and 229 functions, as well as 131 man pages. Cargo won't install them, and it's not a small "just one additional file" task to do that.

Let people in on what we're working with here instead of hinting at it.